File: hdf-converters/shared/go/schemacorpus.go
package shared

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"path/filepath"
	"testing"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	testhdf "github.com/mitre/hdf-libs/hdf-schema/testhdf/go"
	"github.com/stretchr/testify/require"
)

// CorpusContract is the obligation an exporter owes for one corpus input.
//
// A single "HDF-invalid input must be rejected" rule conflated two different
// obligations. A document whose TOP-LEVEL shape is wrong is not the document
// type it claims to be, and refusing it is the only honest outcome. A document
// that is merely wrong in NESTED content is something no converter validates —
// the shared structural guard is deliberately top-level only, and full
// per-conversion schema validation was never the design — so demanding rejection
// there asserted something no converter satisfies, which forced per-converter
// exemptions. Exemptions that accumulate quietly are how a corpus stops meaning
// anything, so the distinction belongs to the case, not to each converter.
type CorpusContract int

const (
	// MustConvert marks sparse but schema-valid HDF: the exporter must convert
	// it, and the output must satisfy the target schema. Refusing legal input is
	// as much a defect as emitting an invalid document for it.
	MustConvert CorpusContract = iota

	// MustReject marks input whose top-level shape the HDF schema rejects — a
	// missing, null, or wrong-typed required collection, or a document that is
	// not an object at all. The exporter must return an error.
	MustReject

	// MustNotCorrupt marks input the HDF schema rejects only in nested content.
	// Either outcome is acceptable — converting it is tolerant, refusing it is
	// strict — but if the exporter does convert, the output must still satisfy
	// the target schema. This is the weaker contract that nested cases can
	// actually be held to, and it still catches the two real defects this class
	// has found: a converter that panics, and one that emits an invalid document.
	MustNotCorrupt
)

// String renders the contract for failure output.
func (c CorpusContract) String() string {
	switch c {
	case MustConvert:
		return "MustConvert"
	case MustReject:
		return "MustReject"
	case MustNotCorrupt:
		return "MustNotCorrupt"
	}
	return "unknown"
}

// CorpusCase is one adversarial exporter input. Every shipped defect this corpus
// exists to catch sat in one of the three contracts above, and a test suite that
// only feeds fully-populated fixtures exercises none of them.
type CorpusCase struct {
	Name string
	// Input is the raw exporter input.
	Input []byte
	// Contract is the obligation the exporter owes for this input. Derived
	// observably rather than asserted: TestCorpusContracts_AreDerivableFromTheSchemaAndGuard
	// checks each case against the real HDF schema AND the shared top-level
	// guard, so a case cannot be quietly reclassified to make a converter pass.
	Contract CorpusContract
	// Why records what the case is probing, surfaced in failure output.
	Why string
}

// mustJSON marshals a builder-produced document. A build-time marshal failure is
// a corpus bug, not a converter bug, so it panics rather than taking a *testing.T
// through every corpus constructor.
func mustJSON(v any) []byte {
	b, err := json.Marshal(v)
	if err != nil {
		panic("schemacorpus: marshal corpus case: " + err.Error())
	}
	return b
}

// withTimestamp stamps the deterministic builder time onto a document so that
// cases probing some other absent field are not silently also probing a missing
// timestamp. Fixed, never wall-clock, per the repo timestamp convention.
func withTimestamp(d hdf.HDFResults) hdf.HDFResults {
	ts := testhdf.DefaultStartTime
	d.Timestamp = &ts
	return d
}

// ResultsCorpus returns the adversarial HDF Results inputs every exporter that
// consumes HDF Results should survive.
//
// MustConvert cases are built with the testhdf builder wherever it can express
// the case. The
// one exception is zero-baselines: Go's variadic testhdf.Doc() leaves Baselines
// nil, and HDFResults.Baselines carries no omitempty, so it marshals to null
// rather than []. It is therefore stated as a typed struct literal — never hand
// written JSON. (The TS peer has no such problem: a JS rest parameter yields [],
// so schema-corpus.ts uses testhdf.doc() directly. The asymmetry is real, not an
// oversight on either side.)
//
// Each MustConvert case isolates exactly one absent field, so a failure names the
// cause. Two cases differing only in a field neither exercises would double the
// runtime and halve the signal.
func ResultsCorpus() []CorpusCase {
	gen := &hdf.Generator{Name: "testhdf", Version: "0.0.0"}

	return []CorpusCase{
		// --- MustConvert: sparse but schema-valid HDF.
		{
			Name:     "zero-baselines",
			Input:    mustJSON(hdf.HDFResults{Baselines: []hdf.EvaluatedBaseline{}, Generator: gen}),
			Contract: MustConvert,
			Why:      "baselines has no minItems, so an assessment that evaluated nothing is legal HDF",
		},
		{
			Name: "no-timestamp",
			// Everything else populated, so a failure here is unambiguously the timestamp.
			Input: mustJSON(testhdf.Results(testhdf.Req("V-1",
				testhdf.Title("t"), testhdf.Severity("medium"), testhdf.Code("c")))),
			Contract: MustConvert,
			Why:      "timestamp is optional in HDF but feeds target fields that are required (XCCDF end-time)",
		},
		{
			Name: "requirement-without-title",
			Input: mustJSON(withTimestamp(testhdf.Results(testhdf.Req("V-1",
				testhdf.Severity("medium"), testhdf.Code("c"))))),
			Contract: MustConvert,
			Why:      "title is optional in HDF but backs target title fields that are often minLength-constrained",
		},
		{
			Name: "requirement-without-code",
			Input: mustJSON(withTimestamp(testhdf.Results(testhdf.Req("V-1",
				testhdf.Title("t"), testhdf.Severity("medium"))))),
			Contract: MustConvert,
			Why:      "absent code must omit the check element entirely, not emit an empty one",
		},
		{
			Name: "requirement-with-non-token-id",
			// Every other MustConvert case uses an id that happens to satisfy
			// OSCAL's token pattern, which is why an exporter copying ids into a
			// token-typed field passed the corpus while failing on 46% of the
			// converters' fixture ids. A package-style id is the commonest
			// non-token shape in real data.
			Input: mustJSON(withTimestamp(testhdf.Results(testhdf.Req("CVE-2018-25032/ruby:nokogiri/1.10.9",
				testhdf.Title("t"), testhdf.Severity("medium"), testhdf.Code("c"))))),
			Contract: MustConvert,
			Why:      "requirement ids are whatever the source tool numbers rules with; target formats often constrain them lexically",
		},
		{
			Name: "requirement-without-severity",
			Input: mustJSON(withTimestamp(testhdf.Results(testhdf.Req("V-1",
				testhdf.Title("t"), testhdf.Code("c"))))),
			Contract: MustConvert,
			Why:      "severity is optional in HDF but target formats constrain it to a fixed vocabulary",
		},

		// --- MustReject: the top-level shape is wrong, so this is not a results
		// document at all. MustNotCorrupt: the top level is fine and only nested
		// content is invalid, which no converter validates.
		{
			Name:     "baselines-missing",
			Input:    []byte(`{"generator":{"name":"t","version":"0.0.0"}}`),
			Contract: MustReject,
			Why:      "baselines is the one required top-level field",
		},
		{
			Name:     "baselines-null",
			Input:    []byte(`{"baselines":null}`),
			Contract: MustReject,
			Why:      "a nil slice marshals to null, so an upstream producer with this bug must be rejected",
		},
		{
			Name:     "baselines-wrong-type",
			Input:    []byte(`{"baselines":"not-an-array"}`),
			Contract: MustReject,
			Why:      "a typed decode can coerce or zero-fill where a structural guard must reject",
		},
		{
			Name:     "baseline-empty-requirements",
			Input:    []byte(`{"baselines":[{"name":"b","requirements":[]}]}`),
			Contract: MustNotCorrupt,
			Why:      "requirements has minItems 1; exporters that map it unguarded emit empty container elements",
		},
		{
			Name:     "requirement-empty-results",
			Input:    []byte(`{"baselines":[{"name":"b","requirements":[{"id":"V-1","impact":0,"tags":{},"descriptions":[{"label":"default","data":"d"}],"results":[]}]}]}`),
			Contract: MustNotCorrupt,
			Why:      "results has minItems 1; exporters that index results[0] unguarded panic",
		},
		{
			Name:     "requirement-missing-id",
			Input:    []byte(`{"baselines":[{"name":"b","requirements":[{"impact":0,"tags":{},"descriptions":[{"label":"default","data":"d"}],"results":[{"status":"passed","codeDesc":"c","startTime":"2020-01-01T00:00:00Z"}]}]}]}`),
			Contract: MustNotCorrupt,
			Why:      "id is required; absent it, exporters emit empty-string identifiers into required target fields",
		},
		{
			Name:     "top-level-array",
			Input:    []byte(`[]`),
			Contract: MustReject,
			Why:      "the one degenerate shape most exporters already reject — pins that they keep doing so",
		},
	}
}

// AmendmentsCorpus returns the adversarial HDF Amendments inputs every exporter
// that consumes HDF Amendments should survive.
func AmendmentsCorpus() []CorpusCase {
	const cve = "CVE-2021-44228"

	return []CorpusCase{
		// --- MustConvert: sparse but schema-valid HDF Amendments.
		{
			Name: "override-empty-reason",
			Input: mustJSON(testhdf.Amendments("a",
				testhdf.Override(hdf.OverrideTypeWaiver, cve,
					testhdf.OverrideStatus(hdf.Failed), testhdf.OverrideReason("")))),
			Contract: MustConvert,
			Why:      "reason carries no minLength, but backs target fields that require non-empty text",
		},
		{
			Name: "override-no-milestones",
			Input: mustJSON(testhdf.Amendments("a",
				testhdf.Override(hdf.OverrideTypeWaiver, cve,
					testhdf.OverrideStatus(hdf.Failed), testhdf.OverrideReason("accepted")))),
			Contract: MustConvert,
			Why:      "milestones are optional; without them remediation text must still be derivable",
		},
		{
			Name:     "evidence-without-description",
			Input:    amendmentsWithBareEvidence(cve),
			Contract: MustConvert,
			Why:      "evidence description is optional but backs CSAF references[].summary, which is minLength 1",
		},

		// --- MustReject: the top-level shape is wrong.
		{
			Name:     "overrides-missing",
			Input:    []byte(`{"name":"a"}`),
			Contract: MustReject,
			Why:      "overrides is required alongside name",
		},
		{
			Name:     "overrides-empty",
			Input:    []byte(`{"name":"a","overrides":[]}`),
			Contract: MustReject,
			Why:      "overrides has minItems 1, so an amendments document that amends nothing is not convertible",
		},
		{
			Name:     "top-level-array",
			Input:    []byte(`[]`),
			Contract: MustReject,
			Why:      "pins that the structural guard rejects a non-object document",
		},
	}
}

// amendmentsWithBareEvidence builds an override carrying url evidence with no
// description. testhdf.Override exposes no evidence option, so the override is
// finished by hand here rather than hand-writing the whole document as JSON —
// keeping the builder's schema-valid scaffolding (appliedAt, expiresAt, identity)
// which is exactly the part that is tedious and easy to get wrong.
func amendmentsWithBareEvidence(reqID string) []byte {
	doc := testhdf.Amendments("a",
		testhdf.Override(hdf.OverrideTypeWaiver, reqID,
			testhdf.OverrideStatus(hdf.Failed), testhdf.OverrideReason("accepted")))
	doc.Overrides[0].Evidence = []hdf.Evidence{{
		Type: hdf.URL,
		Data: "https://example.com/advisory",
	}}
	return mustJSON(doc)
}

// CorpusConvertFn is an exporter entry point taking raw HDF bytes and returning
// its rendered output.
type CorpusConvertFn func(input []byte) ([]byte, error)

// DocumentValidator is the contract the corpus needs from a target-format
// validator: report a document's violations without failing a test. Both the
// JSON Schema validator and the XSD one satisfy it, so a converter whose target
// is an XSD (XCCDF) reuses the same corpus as the JSON-schema converters.
type DocumentValidator interface {
	Validate(doc []byte) error
}

// CheckCase applies the corpus contract to a single case and returns the reason
// it failed, or "" when it passes.
//
// The contract lives here, as a pure function, rather than inline in the
// t.Run loop: assertions buried in a subtest closure cannot be exercised without
// a real *testing.T, so the runner's own logic would go untested — and it was
// exactly an untested branch (a panic satisfying a rejection contract) that
// shipped broken.
//
// A panic fails EVERY contract and is checked before the contract switch. A
// panic is a crash, not a rejection, so letting it satisfy MustReject or
// MustNotCorrupt would green-light the precise defect those contracts exist to
// catch (an unguarded results[0] index is the live example).
func CheckCase(v DocumentValidator, c CorpusCase, convert CorpusConvertFn) string {
	out, err := convertNoPanic(convert, c.Input)

	var panicked *PanicError
	if errors.As(err, &panicked) {
		return fmt.Sprintf("%s: %v (%s)", c.Name, panicked, c.Why)
	}

	switch c.Contract {
	case MustConvert:
		if err != nil {
			return fmt.Sprintf("%s: schema-valid HDF must convert (%s): %v", c.Name, c.Why, err)
		}
		if verr := v.Validate(out); verr != nil {
			return fmt.Sprintf("%s: output does not satisfy the target schema (%s):\n%v", c.Name, c.Why, verr)
		}
		return ""

	case MustReject:
		if err == nil {
			return fmt.Sprintf("%s: input whose top-level shape HDF rejects must not be converted (%s)", c.Name, c.Why)
		}
		return ""

	case MustNotCorrupt:
		// Rejecting is fine: no converter validates nested content, so tolerance
		// and strictness are both defensible. What is never acceptable is
		// converting it into a document the target schema rejects.
		if err != nil {
			return ""
		}
		if verr := v.Validate(out); verr != nil {
			return fmt.Sprintf("%s: converted nested-invalid HDF into output the target schema rejects (%s):\n%v", c.Name, c.Why, verr)
		}
		return ""
	}
	return fmt.Sprintf("%s: unknown contract %v", c.Name, c.Contract)
}

// ValidateCorpus reports why a corpus cannot be meaningfully run, or nil when it
// can. Pure, like CheckCase, so the guard itself is testable — an assertion that
// only exists inside a t.Run closure cannot be exercised without failing the test
// that exercises it.
func ValidateCorpus(cases []CorpusCase) error {
	if len(cases) == 0 {
		return errors.New("corpus is empty — the run would pass vacuously")
	}
	return nil
}

// RunSchemaCorpus asserts every corpus contract against one exporter: see
// CorpusContract for what each obliges. Converters opt in with a single call, so
// the corpus has one definition rather than a copy per converter.
func RunSchemaCorpus(t *testing.T, v DocumentValidator, cases []CorpusCase, convert CorpusConvertFn) {
	t.Helper()
	require.NoError(t, ValidateCorpus(cases))

	for _, c := range cases {
		t.Run(c.Name, func(t *testing.T) {
			if msg := CheckCase(v, c, convert); msg != "" {
				t.Fatal(msg)
			}
		})
	}
}

// convertNoPanic turns a converter panic into an error so a crash is reported as
// a failing case rather than aborting the suite.
func convertNoPanic(convert CorpusConvertFn, input []byte) (out []byte, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = &PanicError{Value: r}
			out = nil
		}
	}()
	return convert(input)
}

// PanicError reports that a converter panicked. It is a distinct type, not a
// plain error, so CheckCase can tell a crash apart from a deliberate rejection:
// a rejection contract is satisfied by an error, and without this distinction a
// panicking converter would pass the very contract meant to catch it.
type PanicError struct{ Value any }

func (e *PanicError) Error() string {
	return "converter panicked (a panic is never an acceptable rejection): " +
		toString(e.Value)
}

func toString(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	if e, ok := v.(error); ok {
		return e.Error()
	}
	b, err := json.Marshal(v)
	if err != nil {
		return "unprintable panic value"
	}
	return string(b)
}

// --- Cross-language parity ---------------------------------------------------

// CanonicalJSON re-serializes a document with map keys sorted, HTML escaping
// disabled, and negative zero normalized, so the same values produce the same
// bytes in Go and TypeScript for every value the corpus can express.
//
// Raw serialization cannot be compared across the two: Go marshals struct fields
// in declaration order while TS uses insertion order, Go escapes <, > and & by
// default while JSON.stringify does not, and Go preserves -0 on a float64 where
// JSON.stringify renders it 0. All three are language artifacts, not meaningful
// differences — normalizing them lets the corpora be asserted byte-equal rather
// than merely assumed equivalent.
//
// Known limit: a lone UTF-16 surrogate still differs (Go substitutes U+FFFD, JS
// preserves the escape). No corpus case can produce one, and one appearing would
// fail the golden loudly rather than pass silently, so it is left unhandled
// rather than carrying speculative code.
func CanonicalJSON(raw []byte) ([]byte, error) {
	var v any
	if err := json.Unmarshal(raw, &v); err != nil {
		return nil, fmt.Errorf("canonical json: %w", err)
	}
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(normalizeNegativeZero(v)); err != nil {
		return nil, fmt.Errorf("canonical json: %w", err)
	}
	// Encode appends a newline; drop it so the value is the whole payload.
	return bytes.TrimRight(buf.Bytes(), "\n"), nil
}

// normalizeNegativeZero rewrites -0 to 0. encoding/json emits "-0" for a
// negative-zero float64 while JSON.stringify emits "0"; since the two are equal
// numerically, preserving the sign would fail the parity golden over a
// difference no consumer can observe.
func normalizeNegativeZero(v any) any {
	switch t := v.(type) {
	case float64:
		if t == 0 {
			return 0.0
		}
		return t
	case []any:
		for i, e := range t {
			t[i] = normalizeNegativeZero(e)
		}
		return t
	case map[string]any:
		for k, e := range t {
			t[k] = normalizeNegativeZero(e)
		}
		return t
	}
	return v
}

// CorpusGoldenEntry is one case as recorded in the cross-language golden.
type CorpusGoldenEntry struct {
	Name string `json:"name"`
	// Contract is recorded so a reclassification on one side alone is caught,
	// not just a renamed or reordered case. Stored as its name rather than its
	// ordinal so the golden stays readable and reordering the constants cannot
	// silently change what it asserts.
	Contract string `json:"contract"`
	// Input is the canonicalized case input.
	Input string `json:"input"`
}

// CorpusGolden is the checked-in cross-language contract: both the Go and the
// TypeScript corpus must reproduce it exactly. Go owns regeneration (go test
// -update); TypeScript only verifies, so neither side can quietly redefine the
// shared corpus to match itself.
type CorpusGolden struct {
	Results    []CorpusGoldenEntry `json:"results"`
	Amendments []CorpusGoldenEntry `json:"amendments"`
}

// CorpusGoldenPath is the golden's location, resolved from this file so it works
// regardless of the package under test.
func CorpusGoldenPath() string {
	return filepath.Join(getSharedDir(), "..", "corpus-golden.json")
}

// BuildCorpusGolden renders the current corpus in golden form.
func BuildCorpusGolden(t *testing.T) CorpusGolden {
	t.Helper()
	return CorpusGolden{
		Results:    goldenEntries(t, ResultsCorpus()),
		Amendments: goldenEntries(t, AmendmentsCorpus()),
	}
}

func goldenEntries(t *testing.T, cases []CorpusCase) []CorpusGoldenEntry {
	t.Helper()
	out := make([]CorpusGoldenEntry, 0, len(cases))
	for _, c := range cases {
		canon, err := CanonicalJSON(c.Input)
		require.NoError(t, err, "canonicalize corpus case %s", c.Name)
		out = append(out, CorpusGoldenEntry{Name: c.Name, Contract: c.Contract.String(), Input: string(canon)})
	}
	return out
}

File: hdf-converters/shared/typescript/schema-corpus.ts
/**
 * Adversarial exporter inputs, shared across every converter's schema tests.
 *
 * The TS peer of shared/go/schemacorpus.go. The two corpora are asserted
 * byte-equal (after canonicalization) against a Go-generated golden, so a case
 * added, renamed, retiered, or altered on one side alone fails the build rather
 * than silently going uncovered on the other.
 */
import * as testhdf from '@mitre/hdf-schema/testhdf';
import { EvidenceType, type HDFResults } from '@mitre/hdf-schema';
import type { ValidateFunction } from 'ajv';
import { schemaErrors } from './schema-validation.js';

/**
 * One adversarial exporter input.
 *
 * `hdfValid` splits the corpus into the two contracts an exporter owes, which
 * are opposites and were previously conflated: a sparse-but-legal document must
 * convert into schema-valid output, while a document HDF itself rejects must
 * produce an error — never a crash, and never a success carrying an invalid
 * document. Every shipped defect this corpus exists to catch sat in one bucket
 * or the other, and a suite that only feeds fully-populated fixtures exercises
 * neither.
 */
export interface CorpusCase {
  name: string;
  input: string;
  /**
   * Whether `input` satisfies the HDF source schema, and so which contract
   * applies. Asserted against the real schema by the corpus tests, so it cannot
   * drift from reality.
   */
  hdfValid: boolean;
  /** What the case is probing, surfaced in failure output. */
  why: string;
}

const CVE = 'CVE-2021-44228';

/** The builder's fixed result time, reused so corpus inputs never touch the wall clock. */
const DEFAULT_TIME = '2020-01-01T00:00:00Z';

/**
 * Stamps the deterministic builder time onto a document so a case probing some
 * other absent field is not silently also probing a missing timestamp.
 *
 * Returns a JSON payload rather than an HDFResults, deliberately: the field is
 * typed Date, and a Date stringifies with milliseconds ("...T00:00:00.000Z")
 * while Go emits the trimmed form. That is a real value difference, which
 * canonicalization would NOT paper over — so the trimmed string is assigned here
 * to keep the corpora byte-equal once canonicalized.
 */
function withTimestamp(d: HDFResults): Record<string, unknown> {
  return { ...d, timestamp: DEFAULT_TIME };
}

/**
 * Adversarial HDF Results inputs every exporter consuming HDF Results should
 * survive. Every tier-A case is built with the testhdf builder, including
 * `zero-baselines` — a JS rest parameter yields [], so `doc()` emits
 * `"baselines": []`. (Go's peer cannot do this: a variadic leaves the slice nil
 * and HDFResults.Baselines has no omitempty, so it marshals to null. That one
 * case is a typed struct literal there. The asymmetry is real, not an oversight.)
 *
 * Each tier-A case isolates exactly one absent field, so a failure names the
 * cause rather than leaving two candidates.
 */
export function resultsCorpus(): CorpusCase[] {
  return [
    // --- Tier A: sparse but schema-valid HDF. Output must satisfy the target schema.
    {
      name: 'zero-baselines',
      input: JSON.stringify(testhdf.doc()),
      hdfValid: true,
      why: 'baselines has no minItems, so an assessment that evaluated nothing is legal HDF',
    },
    {
      // Everything else populated, so a failure here is unambiguously the timestamp.
      name: 'no-timestamp',
      input: JSON.stringify(
        testhdf.results(testhdf.req('V-1', { title: 't', severity: 'medium', code: 'c' })),
      ),
      hdfValid: true,
      why: 'timestamp is optional in HDF but feeds target fields that are required (XCCDF end-time)',
    },
    {
      name: 'requirement-without-title',
      input: JSON.stringify(
        withTimestamp(testhdf.results(testhdf.req('V-1', { severity: 'medium', code: 'c' }))),
      ),
      hdfValid: true,
      why: 'title is optional in HDF but backs target title fields that are often minLength-constrained',
    },
    {
      name: 'requirement-without-code',
      input: JSON.stringify(
        withTimestamp(testhdf.results(testhdf.req('V-1', { title: 't', severity: 'medium' }))),
      ),
      hdfValid: true,
      why: 'absent code must omit the check element entirely, not emit an empty one',
    },
    {
      name: 'requirement-without-severity',
      input: JSON.stringify(
        withTimestamp(testhdf.results(testhdf.req('V-1', { title: 't', code: 'c' }))),
      ),
      hdfValid: true,
      why: 'severity is optional in HDF but target formats constrain it to a fixed vocabulary',
    },

    // --- Tier B: HDF rejects these. The exporter must error, not crash or fabricate.
    {
      name: 'baselines-missing',
      input: '{"generator":{"name":"t","version":"0.0.0"}}',
      hdfValid: false,
      why: 'baselines is the one required top-level field',
    },
    {
      name: 'baselines-null',
      input: '{"baselines":null}',
      hdfValid: false,
      why: 'a nil slice marshals to null, so an upstream producer with this bug must be rejected',
    },
    {
      name: 'baselines-wrong-type',
      input: '{"baselines":"not-an-array"}',
      hdfValid: false,
      why: 'a typed decode can coerce or zero-fill where a structural guard must reject',
    },
    {
      name: 'baseline-empty-requirements',
      input: '{"baselines":[{"name":"b","requirements":[]}]}',
      hdfValid: false,
      why: 'requirements has minItems 1; exporters that map it unguarded emit empty container elements',
    },
    {
      name: 'requirement-empty-results',
      input:
        '{"baselines":[{"name":"b","requirements":[{"id":"V-1","impact":0,"tags":{},"descriptions":[{"label":"default","data":"d"}],"results":[]}]}]}',
      hdfValid: false,
      why: 'results has minItems 1; exporters that index results[0] unguarded crash',
    },
    {
      name: 'requirement-missing-id',
      input:
        '{"baselines":[{"name":"b","requirements":[{"impact":0,"tags":{},"descriptions":[{"label":"default","data":"d"}],"results":[{"status":"passed","codeDesc":"c","startTime":"2020-01-01T00:00:00Z"}]}]}]}',
      hdfValid: false,
      why: 'id is required; absent it, exporters emit empty-string identifiers into required target fields',
    },
    {
      name: 'top-level-array',
      input: '[]',
      hdfValid: false,
      why: 'the one degenerate shape most exporters already reject — pins that they keep doing so',
    },
  ];
}

/**
 * Adversarial HDF Amendments inputs every exporter consuming HDF Amendments
 * should survive. Every tier-A override carries a status because the amendments
 * schema requires an override to declare either `status` or `impact`.
 */
export function amendmentsCorpus(): CorpusCase[] {
  return [
    // --- Tier A: sparse but schema-valid HDF Amendments.
    {
      name: 'override-empty-reason',
      input: JSON.stringify(
        testhdf.amendments('a', testhdf.override('waiver', CVE, { status: 'failed', reason: '' })),
      ),
      hdfValid: true,
      why: 'reason carries no minLength, but backs target fields that require non-empty text',
    },
    {
      name: 'override-no-milestones',
      input: JSON.stringify(
        testhdf.amendments(
          'a',
          testhdf.override('waiver', CVE, { status: 'failed', reason: 'accepted' }),
        ),
      ),
      hdfValid: true,
      why: 'milestones are optional; without them remediation text must still be derivable',
    },
    {
      name: 'evidence-without-description',
      input: amendmentsWithBareEvidence(),
      hdfValid: true,
      why: 'evidence description is optional but backs CSAF references[].summary, which is minLength 1',
    },

    // --- Tier B: HDF rejects these.
    {
      name: 'overrides-missing',
      input: '{"name":"a"}',
      hdfValid: false,
      why: 'overrides is required alongside name',
    },
    {
      name: 'overrides-empty',
      input: '{"name":"a","overrides":[]}',
      hdfValid: false,
      why: 'overrides has minItems 1, so an amendments document that amends nothing is not convertible',
    },
    {
      name: 'top-level-array',
      input: '[]',
      hdfValid: false,
      why: 'pins that the structural guard rejects a non-object document',
    },
  ];
}

/**
 * Builds an override carrying url evidence with no description. testhdf.override
 * exposes no evidence option, so the override is finished by hand here rather
 * than hand-writing the whole document — keeping the builder's schema-valid
 * scaffolding (appliedAt, expiresAt, identity), the tedious part to get right.
 */
function amendmentsWithBareEvidence(): string {
  const override = testhdf.override('waiver', CVE, { status: 'failed', reason: 'accepted' });
  override.evidence = [{ type: EvidenceType.URL, data: 'https://example.com/advisory' }];
  return JSON.stringify(testhdf.amendments('a', override));
}

/** An exporter entry point taking raw HDF text and returning its rendered output. */
export type CorpusConvertFn = (input: string) => Promise<unknown> | unknown;

/**
 * Apply the corpus contract to a single case, returning why it failed or null
 * when it passes.
 *
 * The contract is a pure function, not assertions inside a test closure, so the
 * runner's own logic is testable — an untested branch is what let the Go peer
 * ship a bug where a crash silently satisfied the tier-B contract.
 *
 * A crash fails BOTH tiers and is checked before the tier split. JS has no
 * panic/error distinction, so a crash is identified by what was thrown: any
 * built-in runtime fault (an unguarded index or property access, a bad argument,
 * an escaped JSON.parse failure) or a thrown non-Error, which gives a caller
 * nothing to act on. A deliberate rejection is a plain Error or a converter's own
 * subclass. Letting a runtime fault satisfy tier B would green-light exactly the
 * defect that tier exists to catch.
 */
export async function checkCase(
  validate: ValidateFunction,
  c: CorpusCase,
  convert: CorpusConvertFn,
): Promise<string | null> {
  let out: unknown;
  let threw: unknown;
  let didThrow = false;
  try {
    out = await convert(c.input);
  } catch (e) {
    threw = e;
    didThrow = true;
  }

  if (didThrow && isRuntimeFault(threw)) {
    return `${c.name}: converter crashed with ${describeThrown(threw)} — a crash is never an acceptable rejection (${c.why})`;
  }

  if (c.hdfValid) {
    if (didThrow) {
      return `${c.name}: schema-valid HDF must convert (${c.why}): ${String(threw)}`;
    }
    let parsed: unknown;
    try {
      parsed = parseIfString(out);
    } catch (e) {
      return `${c.name}: output is not parseable JSON (${c.why}): ${String(e)}`;
    }
    const errors = schemaErrors(validate, parsed);
    if (errors !== null) {
      return `${c.name}: output does not satisfy the target schema (${c.why}):\n${errors}`;
    }
    return null;
  }
  if (!didThrow) {
    return `${c.name}: HDF-invalid input must be rejected, not converted (${c.why})`;
  }
  return null;
}

/**
 * Assert both corpus contracts against one exporter: tier-A cases must convert
 * and satisfy the schema, tier-B cases must be rejected.
 *
 * Converters opt in with a single call, so the corpus has one definition rather
 * than a copy per converter. Every failing case is collected and reported in one
 * throw, so a red run names all of them rather than one per re-run. (Go's peer
 * needs no equivalent: each case is its own subtest and already reports
 * independently.)
 */
export async function runSchemaCorpus(
  validate: ValidateFunction,
  cases: CorpusCase[],
  convert: CorpusConvertFn,
): Promise<void> {
  if (cases.length === 0) throw new Error('corpus is empty — the run would pass vacuously');

  const failures: string[] = [];
  for (const c of cases) {
    const failure = await checkCase(validate, c, convert);
    if (failure !== null) failures.push(failure);
  }
  if (failures.length > 0) throw new Error(failures.join('\n\n'));
}

/**
 * Built-in error types that signal a fault in the converter rather than a
 * decision by it. A converter rejecting bad input throws a plain Error; one that
 * hits any of these has lost control of its own execution.
 */
const RUNTIME_FAULTS = new Set([
  'TypeError',
  'RangeError',
  'ReferenceError',
  'SyntaxError',
  'EvalError',
  'URIError',
]);

/** Whether a thrown value represents a crash rather than a deliberate rejection. */
function isRuntimeFault(thrown: unknown): boolean {
  // A thrown non-Error (string, undefined, object literal) carries no message or
  // stack, so it is not a rejection a caller can diagnose either.
  if (!(thrown instanceof Error)) return true;
  return RUNTIME_FAULTS.has(thrown.name);
}

/** Renders a thrown value for a failure message, whatever its type. */
function describeThrown(thrown: unknown): string {
  if (thrown instanceof Error) return `${thrown.name}: ${thrown.message}`;
  return `a non-Error value (${typeof thrown})`;
}

/** Converters return either a JSON string or an object; validate the object form. */
function parseIfString(out: unknown): unknown {
  return typeof out === 'string' ? (JSON.parse(out) as unknown) : out;
}

// --- Cross-language parity ---------------------------------------------------

/**
 * Re-serialize a document with object keys sorted, so the same values produce
 * the same bytes in TypeScript and Go.
 *
 * Raw serialization cannot be compared across the two: Go marshals struct fields
 * in declaration order while TS uses insertion order. Both are language
 * artifacts, not meaningful differences — canonicalizing removes them so the
 * corpora can be asserted byte-equal rather than merely assumed equivalent.
 * (Go additionally disables its default HTML escaping of <, > and & to match
 * JSON.stringify, and normalizes -0 to 0 where JSON.stringify already does.)
 *
 * Known limit: a lone UTF-16 surrogate still differs between the two. No corpus
 * case can produce one, and one appearing would fail the golden loudly rather
 * than pass silently.
 */
export function canonicalJSON(raw: string): string {
  // Go's encoder escapes U+2028/U+2029; JSON.stringify emits them literally.
  // Escaping here keeps the two byte-equal (and the output valid JavaScript,
  // where a bare U+2028 is a line terminator).
  return JSON.stringify(sortValue(JSON.parse(raw)))
    .replace(/\u2028/g, '\\u2028')
    .replace(/\u2029/g, '\\u2029');
}

function sortValue(v: unknown): unknown {
  if (Array.isArray(v)) return v.map(sortValue);
  if (v === null || typeof v !== 'object') return v;
  const src = v as Record<string, unknown>;
  const out: Record<string, unknown> = {};
  for (const k of Object.keys(src).sort()) out[k] = sortValue(src[k]);
  return out;
}

/** One case as recorded in the cross-language golden. */
export interface CorpusGoldenEntry {
  name: string;
  hdfValid: boolean;
  input: string;
}

/** Renders a corpus in golden form for comparison against the checked-in file. */
export function toGoldenEntries(cases: CorpusCase[]): CorpusGoldenEntry[] {
  return cases.map((c) => ({ name: c.name, hdfValid: c.hdfValid, input: canonicalJSON(c.input) }));
}

